File: python/test/wsweepkerrcomp.py
#!/usr/bin/env python3
"""
Compare kernel formula errors vs. kernel width, mirroring matlab/test/wsweepkerrcomp.m.

This script uses erralltypedim.py for the core error computations and calls the
FINUFFT Python bindings to measure kernel support via spreadinterponly for each
kernel formula, producing a figure saved under results/.
"""
from __future__ import annotations

import math
import os
import sys

import matplotlib.pyplot as plt
import numpy as np
import warnings

from erralltypedim import erralltypedim

import finufft


def _count_nonzero_support(du: np.ndarray) -> int:
    """Return the number of nonzero entries in the spreadinterponly output."""
    return int(np.count_nonzero(np.asarray(du) != 0.0))


def main() -> None:
    """Run the kernel-width sweep and save the comparison plot to results/."""
    import argparse

    p = argparse.ArgumentParser(description="Kernel-width sweep")
    p.add_argument("--sigma", type=float, default=2.00, help="upsampling factor (sigma)")
    p.add_argument("--dim", type=int, default=1, choices=[1, 2, 3], help="dimension to test")
    p.add_argument("--M", type=int, default=1000, help="number of nonuniform points (M)")
    p.add_argument("--Ntot", type=int, default=300, help="total number of modes (Ntot)")
    p.add_argument("--ntr", type=int, default=10, help="# transforms to average per tol (ntr)")
    args = p.parse_args()

    prec = "double"
    myrand = np.random.rand
    M = int(args.M)
    dim = int(args.dim)
    Ntot = int(args.Ntot)
    ntr = int(args.ntr)
    isign = +1
    sigma = float(args.sigma)
    tolsperdecade = 8
    tolstep = 10 ** (-1 / tolsperdecade)
    kfnam = ["ES legacy", "ES Beatty", "KB Beatty", "cont-KB Beatty", "cosh-type", "smoothed cont-KB"]
    kfs = list(range(1, len(kfnam) + 1))

    o = {"upsampfac": sigma, "showwarn": False}
    dims = [False, False, False]
    dims[dim - 1] = True
    nkf = len(kfs)

    mintol = 10 * np.finfo(np.float64).eps
    ntols = max(1, math.ceil(math.log(mintol) / math.log(tolstep)))
    ntols = int(ntols)
    tols = tolstep ** np.arange(ntols, dtype=float)

    print(
        f"{dim}D sigma={sigma:.3g}\tprec={prec} M={M} Ntot={Ntot} ntr={ntr} "
        f"ntols={ntols}, kfs: {' '.join(str(k) for k in kfs)}"
    )

    # suppress repeated FINUFFT eps-too-small warnings during sweeps
    warnings.filterwarnings("ignore", message="FINUFFT eps tolerance too small to achieve")

    errs = np.full((nkf, 3, ntols), np.nan, dtype=float)
    ws = np.zeros((nkf, ntols), dtype=int)

    for t, tol in enumerate(tols):
        for i, kf in enumerate(kfs):
            o["spread_kerformula"] = kf
            errvals, info = erralltypedim(
                M, Ntot, ntr, isign, prec, float(tol), o, myrand, dims
            )
            errs[i, :, t] = np.asarray(errvals)[:, dim - 1]

            o_spread = dict(o)
            o_spread["spreadinterponly"] = 1
            du = finufft.nufft1d1(
                np.array([0.0], dtype=float),
                np.array([1.0], dtype=np.complex128),
                n_modes=100,
                eps=tol,
                isign=isign,
                **o_spread,
            )
            ws[i, t] = _count_nonzero_support(du)

    wmax = int(np.max(ws))
    if wmax < 2:
        raise RuntimeError("Spread support never reached w>=2; cannot draw comparison.")

    ekw = np.full((nkf, 3, wmax + 1), np.nan, dtype=float)
    vkw = np.full_like(ekw, np.nan)

    for w in range(2, wmax + 1):
        for i in range(nkf):
            tt = np.flatnonzero(ws[i] == w)
            if tt.size == 0:
                continue
            slice_errs = errs[i, :, tt]
            # slice_errs may be shape (3, n) or (n, 3) or (3,) if n==1.
            if slice_errs.ndim == 1:
                # single sample: assign directly and set zero variance
                ekw[i, :, w] = slice_errs
                vkw[i, :, w] = 0.0
            else:
                # prefer axis=1 if first dim==3, else axis=0
                if slice_errs.shape[0] == 3:
                    mean_vals = np.nanmean(slice_errs, axis=1)
                    var_vals = np.nanvar(slice_errs, axis=1)
                elif slice_errs.shape[1] == 3:
                    mean_vals = np.nanmean(slice_errs, axis=0)
                    var_vals = np.nanvar(slice_errs, axis=0)
                else:
                    # fallback: compute mean along last axis and hope for shape (3,)
                    mean_vals = np.nanmean(slice_errs, axis=1)
                    var_vals = np.nanvar(slice_errs, axis=1)
                ekw[i, :, w] = mean_vals
                vkw[i, :, w] = var_vals

    fig = plt.figure(figsize=(15, 5))
    results_dir = os.path.abspath(os.path.join(os.getcwd(), "results"))
    os.makedirs(results_dir, exist_ok=True)

    w_values = np.arange(2, wmax + 1)
    min_err = np.nanmin(ekw[:, :, 2:])
    max_err = np.nanmax(ekw[:, :, 2:])

    for y in range(3):
        ax = fig.add_subplot(1, 3, y + 1)
        legs = []
        for i in range(nkf):
            means = ekw[i, y, 2:]
            stddev = np.sqrt(vkw[i, y, 2:])
            valid = np.isfinite(means) & np.isfinite(stddev)
            if not np.any(valid):
                continue
            ax.errorbar(
                w_values[valid],
                means[valid],
                yerr=stddev[valid],
                fmt=".-",
                markersize=4,
                linewidth=1,
                elinewidth=0.6,
                capsize=2,
            )
            legs.append(f"kf={kfs[i]}: {kfnam[i]}")
        ax.set_yscale("log")
        ax.set_xlim(2, wmax)
        if np.isfinite(min_err) and np.isfinite(max_err):
            ax.set_ylim(min_err, max_err)
        ax.set_xlabel("w")
        ax.set_ylabel("mean rel err")
        if legs:
            ax.legend(legs)
        ax.set_title(
            f"{dim}D type {y + 1} {prec}, N_tot={Ntot}, σ={sigma}", pad=6
        )
        ax.tick_params(axis="both", which="both", width=0.5, length=4)
        ax.set_facecolor("white")

    plt.tight_layout()
    outfile = os.path.join(
        results_dir, f"wsweepkerrcomp_{dim}D_{prec}_sig{sigma}.png"
    )
    fig.savefig(outfile)
    print(f"Saved {outfile}")


if __name__ == "__main__":
    main()

File: matlab/test/wsweepkerrcomp.m
% Comparison of several kerformulae: error vs widths w, 3 types, fix dim,
% CPU only. One sigma for now.
% This measures the goodness of shape (beta) choice for each type (kf) and w.
% It is unaffected by the logic of choosing w appropriate for a given tol.
% 
% Based on tol sweep from tolsweeptest, using erralltypedim to meas err,
% and extracts w from a spreadinterponly call (ignores tol).
% Barnett 1/16/26. Added err-vs-tol comparison 1/20/26; PSWF 1/23/26.

addpath(fileparts(mfilename('fullpath')))
clear
prec = 'double';  % working precision
myrand = @rand;   % select CPU

M = 1e3;             % # NU pts (several secs for >=1e4)
dim = 1; Ntot = 300; % which dimensionality to test, tot #modes (not too small)
% (weird thing is N small, eg 32, makes KB look better)
%dim = 2; Ntot = 400;  % or try other dims...
                      %dim = 3; Ntot = 1e3;
ntr = 20;           % #transforms to average error over at each tol (was 10)
isign = +1;
sigma = 2;
tolsperdecade = 10;
tolstep = 10 ^ (-1 / tolsperdecade); % multiplicative step in tol, < 1
% following names must match src/finufft_common/kernel.h:
kfnam = {"ES legacy", "ES Beatty", "KB Beatty", "cont-KB Beatty", "cosh-type Bea", "cont cosh Bea", "PSWF Beatty", "PSWF beta-shift", "PSWF Marco"};
kfs = [1 8];       % kernel formulae to test

o.upsampfac = sigma;
%o.debug = 1;
o.showwarn = 0; warning('off','FINUFFT:epsTooSmall');
dims = false(1, 3); dims(dim) = true;  % only test this dim
nkf = numel(kfs);
mintol = 10 * eps(prec);        % stop above eps_mach
ntols = ceil(log(mintol) / log(tolstep));
tols = tolstep.^(0:ntols-1);     % go down from tol = 1
fprintf('%dD sigma=%.3g\tprec=%s M=%d Ntot=%d ntr=%d ntols=%d, kfs:',...
        dim, o.upsampfac, prec, M, Ntot, ntr, ntols);
fprintf(' %d',kfs); fprintf('\n');
errs = nan(nkf, 3, ntols);     % for 3 types (just 1D for now), each tol
toloks = true(1,ntols);        % whether FINUFFT reported warning for tol
ws = zeros(nkf, ntols);        % extracted widths w
for t=1:ntols
  tol = tols(t);
  for i = 1:numel(kfs)  % loop over kernel formulae
    o.spread_kerformula = kfs(i);
    %o.debug = (tol<1e-9 && tol>1e-10);  % *** only to find ns=15 s=1.25 bump :(
    lastwarn('');                  % clean up warnings
    [nineerrs, info] = erralltypedim(M,Ntot,ntr,isign,prec,tol,o,myrand,dims);
    errs(i,:,t) = nineerrs(:,dim);   % extract col from 3x3
    [~,id] = lastwarn; toloks(t) = ~strcmp(id, 'FINUFFT:epsTooSmall');
    % measure w via support of one spread pt to the origin (type-1 only!)...
    oo = o; oo.spreadinterponly = 1;
    if dim==1, du = finufft1d1(0,1,+1,tol,32,oo);   % Nj>=2.ns here
    elseif dim==2, du2 = finufft2d1(0,0,1,+1,tol,32,32,oo); du = sum(du2);
    else, du3 = finufft3d1(0,0,0,1,+1,tol,32,32,32,oo); du = sum(du3,[1 2]); end
    ws(i,t) = sum(du~=0.0);
  end
end

% gather mean errs (ekw) for each kerformula, type, and w...
wmax = max(ws(:)); ekw = nan(nkf,3,wmax); vkw=ekw;
for w = 2:wmax
  for i = 1:numel(kfs)
    tt = find(ws(i,:)==w);   % take mean and var along tol t ind...
    [vkw(i,:,w), ekw(i,:,w)] = var(squeeze(errs(i,:,tt)),0,2);
  end
end

% do the err vs w plot...
figure('name','mean rel err vs w, comparing kernels, for 3 NUFFT types',...
       'position',[200 200 1200 500]);
for y=1:3  % types
  subplot(1,3,y);
  legs = {};
  for i=1:nkf     % kernels. make error bars with +-stddev...
    errorbar(2:wmax, squeeze(ekw(i,y,2:end)), sqrt(squeeze(vkw(i,y,2:end))),'.-','markersize',10);
    set(gca, 'ysc','log'); hold on; xlabel('w'); ylabel('mean rel l2 err');
    legs{i} = sprintf('kf=%d: %s',kfs(i),kfnam{kfs(i)});
  end
  axis([2 wmax min(ekw(:)) max(ekw(:))]);
  legend(legs)
  title(sprintf('%dD type %d %s, N_{tot}=%d, \\sigma=%g',dim,y,prec,Ntot,sigma))
end
print('-dpng',sprintf('results/wsweepkerrcomp_%dD_%s_sig%g.png',dim,prec,sigma))

% print # digits err change between first & 2nd formula (3 cols for types)...
disp('digits changed vs ns & type:'); log10(ekw(2,:,:)./ekw(1,:,:))

% also for kicks plot err vs tol for these kernels...
figure('name','mean rel err vs tol, comparing kernels, for 3 NUFFT types',...
       'position',[200 200 1200 500]);
for y=1:3  % types
  subplot(1,3,y);
  legs = {};
  symb = '+.ox*sd';
  tt = tols(toloks);            % plot only the non-warning tol domain
  for i=1:nkf     % kernels
    loglog(tt, squeeze(errs(i,y,toloks)), symb(i)); % 'markersize',10);
    hold on; xlabel('\epsilon (user tol)'); ylabel('mean rel l2 err');
    legs{i} = sprintf('kf=%d: %s',kfs(i),kfnam{kfs(i)});
  end
  plot(tt,tt,'-'); legs{nkf+1} = '\epsilon';
  axis([min(tt) max(tols) min(tt) 1.0]);
  legend(legs,'location','nw')
  title(sprintf('%dD type %d %s, N_{tot}=%d, \\sigma=%g',dim,y,prec,Ntot,sigma))
end
print('-dpng',sprintf('results/tolsweepkerrcomp_%dD_%s_sig%g.png',dim,prec,sigma))

